### PR TITLE
[networks] Use NAT information to generate `ByteKey`

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

Use NAT information to generate `ByteKeys` which are used as unique identifiers of TCP connections for the purposes of storage and aggregation inside `network.State`.

This came up when investigating orphan HTTP stats that couldn't be joined to their TCP connection stats.

In the context of a server with high connection churn, there is a chance that 2 or more ephemeral connections sharing the same source and destination addresses (due to port-reuse) will have different NAT translations associated to each.
This PR ensures that these connections ensures that these connections won't be inadvertently merged, since they do differ in their post NAT addresses.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/connection/kprobe/tcp_close_consumer.go [hunk 1]
```diff
@@ -119,6 +128,20 @@ func (c *tcpCloseConsumer) Start(callback func([]network.ConnectionStats)) {
 				c.batchManager.GetPendingConns(oneTimeBuffer)
 				callback(oneTimeBuffer.Connections())
 				close(request)
+
+				closedCount += oneTimeBuffer.Len()
+				now := time.Now()
+				elapsed := now.Sub(then)
+				then = now
+				log.Debugf(
+					"tcp close summary: closed_count=%d elapsed=%s closed_rate=%.2f/s lost_count=%d",
```
Comment (p-lambert) [inline pkg/network/tracer/connection/kprobe/tcp_close_consumer.go:137]: This is not directly related to the fix, but I've found this information extremely useful during multiple investigations, so I think we should add this log entry.

**Comment by brycekahle** [comment]
Could you now have two connections with different non-NAT addresses, but the same translation (because they used it at different times) aggregating? It feels like to avoid both sides of the problem, the key needs to include both pre-NAT and post-NAT?

**Comment by p-lambert** [comment]
> Could you now have two connections with different non-NAT addresses, but the same translation (because they used it at different times) aggregating? It feels like to avoid both sides of the problem, the key needs to include both pre-NAT and post-NAT?

@brycekahle while in theory possible that is possible I think that represents a departure from what we currently do everywhere else (in the sense that we use post-NAT information as the _defacto_ address, and not as part of a composite key like (pre-nat, post-nat) to identify connections.

For example, both DNS and HTTP data have only post-NAT IP addresses so we wouldn't be able to determine how to join them to their corresponding `ConnectionStats` objects.

File: pkg/network/event_common.go [hunk 2]
```diff
@@ -260,27 +260,14 @@ func (c ConnectionStats) IsExpired(now uint64, timeout uint64) bool {
...
-	n += 8
-
-	// Family (4 bits) + Type (4 bits) = 8 bits
-	buf[n] = uint8(c.Family)<<4 | uint8(c.Type)
-	n++
-
-	n += c.Source.WriteTo(buf[n:]) // 4 or 16 bytes
-	n += c.Dest.WriteTo(buf[n:])   // 4 or 16 bytes
-	return buf[:n], nil
+func (c ConnectionStats) ByteKey(buf []byte) []byte {
```
Comment (hmahmood) [inline pkg/network/event_common.go:269]: nit: I would prefer something like another function `NATByteKey` or `ByteKeyWithNAT` (exported or not).

File: pkg/network/state.go [hunk 3]
```diff
@@ -93,12 +93,16 @@ type telemetry struct {
 
 const minClosedCapacity = 1024
 
+type closedKey string
```
Comment (hmahmood) [inline pkg/network/state.go:96]: Not sure we need this anymore; I'd be ok with removing it and using `string`.

**Review by hmahmood — APPROVED** [review]

File: pkg/network/state.go (hunk 3 above)
Comment (p-lambert) [reply pkg/network/state.go:96]: 👍
